### ISSUE TITLE
Tensorflow inference support

### ISSUE THREAD

**Issue by jkiliani** [leela-zero/leela-zero#1188]
Recently, @mooskagh prepared a Tensorflow implementation for Leela chess (https://github.com/glinscott/leela-chess/pull/262). It is still in an early stage and missing a lot of features, but it gave a very impressive speedup by allowing Tensorflow to utilise CuDNN neural network processing routines.

Could this be an option for Leela Zero to avoid the Nvidia licensing issues (https://github.com/gcp/leela-zero/issues/875) from linking to CuDNN directly? I don't know a lot about software licensing, but my impression from the discussion on Leela chess was that this implementation makes the developers of Leela chess only subject to the Tensorflow license, but not the restrictive CuDNN license. Of course, each user who wanted maximum performance would have to install CuDNN by themselves...

Edit: Compared to @Ttl's solution, the implementation by @mooskagh seems to derive most of its speedup by batching, using virtual loss. The strength loss from the inefficiencies in the search itself seems to be much more than compensated by the calculation speed increase. However, at fixed visits the batching solution is somewhat weaker.

**Comment by jkiliani** [comment]
It doesn't? I was under the impression the liability for the repository owner was the main problem with linking CuDNN directly...

**Comment by gcp** [comment]
>my impression from the discussion on Leela chess was that this implementation makes the developers of Leela chess only subject to the Tensorflow license

The combined work would still be GPLed, so I don't see why this would be true.

I saw that the person in that topic did a complete re-implementation, so I figured they'd have used the BSD license (or similar) to work around the problem, because it would allow you to distribute an .exe that the user then has to link with cuDNN (which they have to obtain), but somewhat to my surprise the new one is GPLed too.

**Comment by jkiliani** [comment]
The developer in question works at Google but did say that he's not an expert in licensing... so would this current implementation likely lead to problems for Leela chess?

**Comment by gcp** [comment]
At a cursory look it looked like an independent re-implementation so they can do whatever they want with it?

**Comment by jkiliani** [comment]
OK, good to know. So the only way we could have something similar for Leela Zero would be a complete reimplementation of the engine under BSD license? There's no way to make this work with the current, GPL binary?

**Comment by gcp** [comment]
This would allow you to produce a .exe that is linked to the cuDNN headers/support libs, which you could then redistribute, but the user wouldn't be able to do anything with it until they get an NVIDIA account and obtain cuDNN themselves.

(So what it saves you is that the user doesn't have to compile the program by themselves if they want to use cuDNN support)

As for "liability of the repository owner", that's more of a problem with the cuDNN license itself, which still has pretty onerous requirements regarding bookkeeping and requires you to snitch etc, and that's independent of the license of Leela (Chess) Zero or any re-implementation.

**Comment by jkiliani** [comment]
Would Tensorflow be considered a support lib under CuDNN? If so, would this not make the license problem a problem of Google instead of whoever used Tensorflow? After all, Tensorflow doesn't have to use CuDNN to somewhat work, or does it?

**Comment by gcp** [comment]
>If so, would this not make the license problem a problem of Google instead of whoever used Tensorflow?

The license is the problem of the person doing the distribution. (The license gives you permissions that copyright law otherwise wouldn't give you).

Google is in no way involved in this.

**Comment by gcp** [comment]
By the way, early in the project someone made a TCP/UDP backend for LZ that interfaced with Theano (IIRC) and used cuDNN. That's another way of avoiding the problem.

**Comment by jkiliani** [comment]
It seems Leela chess will try to talk to Nvidia directly, let's see how that pans out...
